cartography/intel/gcp/crm.py

```diff
@@ -45,14 +69,158 @@ def load_gcp_organizations(neo4j_session, data, gcp_update_tag):
         )
 
 
+def load_gcp_folders(neo4j_session, data, gcp_update_tag):
+    """
+    Ingest the GCP folders to Neo4j
+    :param neo4j_session: The Neo4j session
+    :param data: List of folders; output from crm.get_gcp_folders()
+    :param gcp_update_tag: The timestamp value to set our new Neo4j nodes with
+    :return: Nothing
+    """
+    for folder in data:
+        # Get the correct parent type.
+        # Parents of folders can only be GCPOrganizations or other folders, see
+        # https://cloud.google.com/resource-manager/docs/cloud-platform-resource-hierarchy
+        if folder['parent'].startswith("organizations"):
+            query = "MATCH (parent:GCPOrganization{id:{ParentId}})"
+        elif folder['parent'].startswith("folders"):
+            query = """
+            MERGE (parent:GCPFolder{id:{ParentId}})
+            ON CREATE SET parent.firstseen = timestamp()
+            """
+        query += """
+        MERGE (folder:GCPFolder{id:{FolderName}})
+        ON CREATE SET folder.firstseen = timestamp()
+        SET folder.foldername = {FolderName},
+        folder.displayname = {DisplayName},
+        folder.lifecyclestate = {LifecycleState},
+        folder.lastupdated = {gcp_update_tag}
+        WITH parent, folder
+        MERGE (parent)-[r:RESOURCE]->(folder)
+        ON CREATE SET r.firstseen = timestamp()
+        SET r.lastupdated = {gcp_update_tag}
+        """
+        neo4j_session.run(
+            query,
+            ParentId=folder['parent'],
+            FolderName=folder['name'],
+            DisplayName=folder.get('displayName', None),
+            LifecycleState=folder.get('lifecycleState', None),
+            gcp_update_tag=gcp_update_tag
+        )
+
+
+def load_gcp_projects(neo4j_session, data, gcp_update_tag):
+    """
+    Ingest the GCP projects to Neo4j
+    :param neo4j_session: The Neo4j session
+    :param data: List of GCP projects; output from crm.get_gcp_projects()
+    :param gcp_update_tag: The timestamp value to set our new Neo4j nodes with
+    :return: Nothing
+    """
+    for project in data:
+        if project['parent']['type'] == "organization":
+            query = "MATCH (parent:GCPOrganization{id:{ParentId}})"
+            parentid = f"organizations/{project['parent']['id']}"
+        elif project['parent']['type'] == "folder":
+            query = """
+            MERGE (parent:GCPFolder{id:{ParentId}})
```

Comment:
Do you think it's worth walking up the folder tree and creating a relationship from the organization to _all_ projects it contains, regardless of which folder they're in? I don't think this is something we should add in this PR, but we should think about it for the future IMO.

Reply:
I see your point, since if you're using the GCP API you don't necessarily need to care about a project being in a folder (well, for that point you don't really need to care about a project being under an organization tbh, you just need to know the projectId).  We can think about creating these shortcut relationships, though this is partially mitigated by having everything set as :RESOURCE as in my comment earlier.